sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -188,11 +188,19 @@ class UnivocityParser(
     }
   }
 
+  private val doParse = if (requiredSchema.nonEmpty) {
+    (input: String) => convert(tokenizer.parseLine(input))
+  } else {
+    // If `columnPruning` enabled and partition attributes scanned only,
+    // `schema` gets empty.
+    (_: String) => InternalRow.empty
```

Comment:
It's too long ago and I can't remember the details. Does it mean we still have this count optimization for CSV? does it work in multiline mode?

Reply:
Yes, it does for CSV and for miltiline mode, it executes a different code path.

`UnivocityParser.parseStream` -> `UnivocityParser.convert`